### PR TITLE
Add Annotations, links, time events

### LINKED ISSUE #78: Cannot set attributes on a detached span after the span is attached
Note: the following also applies to adding annotations, links, message events, etc. to spans.

Consider:
```php
$span = Tracer::startSpan(['name' => 'root']);
$scope = Tracer::withSpan($span);
$span->addAttribute('foo', 'bar');
$scope->close();
```

This will work using the `ContextTracer` as the `$span` is a reference to the tracked span. This will not work for the extension as the `$span` is not a reference to the object being held in the extension's memory.

We could resolve this by delegating the addition of attributes to the `TracerInterface` implementation:
```php
interface TracerInterface
{
    ...
    // option 1
    public function addAttribute(Span $span, $key, $value);

    // option 2, $options['span'] would be the span, otherwise assume the current span.
    public function addAttribute($key, $value, $options = []);
}
```
The previous example would look like:
```php
$span = Tracer::startSpan(['name' => 'root']);
$scope = Tracer::withSpan($span);

// option 1
Tracer::addAttribute($span, 'foo', 'bar');

// option 2
Tracer::addAttribute('foo', 'bar', ['span' => $span]);

$scope->close();
```

Option 1 precludes you from being able to add attributes to the current span (common in the implicit span usage). Option 2 gets a bit hairy when applied to something like annotations or links where those models also have optional values.

### PR DESCRIPTION
**Pull request by chingor13**
Fixes #78 

Adds annotations, links, time events to the tracers.

### CODE REVIEW

File: src/Trace/Annotation.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace OpenCensus\Trace;
+
+/**
+ * A class that represents a Annotation resource.
```
Comment (tmatsuo) [inline src/Trace/Annotation.php]: an?

File: src/Trace/TimeEvent.php [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    /**
+     * Set the time for this event.
+     *
+     * @param \DateTimeInterface|int|float $time The time of this event.
+     */
+    public function setTime($time = null)
+    {
+        if (!$when) {
```
Comment (tmatsuo) [inline src/Trace/TimeEvent.php]: `$when` first appears here. Seems like you need to change `$time` to `$when` in the arg list?

File: ext/opencensus_trace_span.c [hunk 3]
```diff
@@ -176,69 +186,86 @@ static PHP_METHOD(OpenCensusTraceSpan, attributes) {
...
 
-    val = zend_read_property(opencensus_trace_span_ce, getThis(), "startTime", sizeof("startTime") - 1, 1, &rv);
+    val = zend_read_property(opencensus_trace_span_ce, getThis(), "links", sizeof("links") - 1, 1, &rv);
 
     RETURN_ZVAL(val, 1, 0);
 }
 
 /**
- * Fetch the end time
+ * Fetch the attributes for the span
```
Comment (tmatsuo) [inline ext/opencensus_trace_span.c]: Wrong comment?

File: src/Trace/Span.php [hunk 4]
```diff
@@ -251,13 +263,67 @@ public function name()
...
+    {
+        foreach ($links as $link) {
+            $this->addLink($link);
+        }
+    }
+
+    /**
+     * Add a link to this span.
+     *
+     * @param TimeEvent $timeEvent
```
Comment (tmatsuo) [inline src/Trace/Span.php]: Wrong `@param`?

File: ext/opencensus_trace.c [hunk 5]
```diff
@@ -119,6 +140,36 @@ PHP_FUNCTION(opencensus_version)
...
+{
+    zend_string *span_id = NULL;
+    opencensus_trace_span_t *span = NULL;
+
+    if (options == NULL) {
+        return NULL;
+    }
+
+    if (span_id = span_id_from_options(Z_ARR_P(options))) {
+        span = (opencensus_trace_span_t *)zend_hash_find_ptr(OPENCENSUS_TRACE_G(spans), span_id);
```
[the remark this answers is not in the document]
Comment (chingor13) [reply ext/opencensus_trace.c]: You mean if null is explicitly passed to the options? Or if no span is provided?

[the remark this answers is not in the document]
Comment (tmatsuo) [reply ext/opencensus_trace.c]: Alright. Maybe it's fine with silent fallback

**Pull request merged**